servant-client: generalize the function to set the request body

kosmikus opened the pull request:
NOTE: This is most probably not ready to be merged yet. It's primarily to aid the discussion in #656.

Rather than hard-coding the `RequestBodyLBS` constructor and be
limited to lazy bytestrings, the new function `setReqBody` just
takes any value of type `RequestBody`.

This has the advantage the we can use streaming request bodies
such as for example constructed by the `streamFile` function in
http-client.

alpmestan commented:
I like that this doesn't change the `ReqBody` instance. I'm wondering however if we could allow users to pick between LBS, streaming etc in an easy way, without having to define a new combinator for request bodies that should be sent using some particular constructor. Otherwise everyone will just end up writing similar combinators just to pick the streaming constructor for example.

Also @phadej is actively thinking about ways to make -server and -client less about wai/http-client, he might want to chime in.

Note: we should obviously find a better name for the new request body setting function, but we're not going to merge this as is anyway and this is a detail, so we can keep that for the end :)

kosmikus commented:
@alpmestan Yes, this all makes sense. However, I think it would be better to disentangle the bigger picture from a small local fix that could even be fully backwards-compatible. This is because AFAIK, the problem @edsko was mentioning is currently a blocker in a project he's working on. So it might be good to have a quick fix and release if there's a small change we can all agree on for now.

phadej commented:
AS said in the issue, I'm ok with throwing `http-client` specific stuff in atm. Elegant but broken solution is worse than not-so-elegant but working.

And this is elegant enough.

alpmestan commented:
OK, I think it indeed makes sense to merge this in master and ship it in the next release. My hope for a way to be able to easily switch from LBS to streaming for the request body is a user-facing API problem and should not block this from being merged.

alpmestan commented:
I'm all for merging this. Unless anyone objects, let's merge this and include it in the upcoming release?

phadej commented:
I don't have rights to merge myself, but go ahead!

kosmikus commented:
@alpmestan Ok. I think we should wait for @edsko to confirm that this does indeed fix his situation, too, and not just the simplified example he created. Also, we should then agree on the actual name(s) being exposed. I chose setReqBody for the new function because it was free and because setRQBody is a naming scheme I cannot completely explain anyway. But having both function, especially without comments, is not very nice. Any better suggestions?

phadej commented:
I'd rename `setRQBody` to `setReqBodyLBS`, and maybe keep the `setRQBody` name for sometime, but deprecated.

kosmikus commented:
Ok, this is close to what I had in mind. So if I gather correctly, the suggestion is to:

- deprecate `setRQBody` and say it has been renamed to `setReqBodyLBS`
- add `setReqBodyLBS` with the same functionality as `setRQBody`
- keep the new `setReqBody` under this name
- add short documentation for both `setReqBody` and `setReqBodyLBS`.

I'll try to amend the PR.

phadej commented:
Feel free to merge after @edsko is verified it actually fixes the problem.

kosmikus commented:
I've implemented the name changes as discussed and also bumped the version and added since-annotations for the new functions (none of the other functions have any, but we can at least start somewhere).

codedmart commented:
LGTM, seems straight forward enough.

alpmestan commented:
Thanks @kosmikus :)
